### DISCUSSION TITLE
Missmatch version between tauri/server vs tauri/js?

### DISCUSSION THREAD

**Discussion by jourdain** [tauri-apps/tauri#5463] [Q&A] [answered]
I'm trying to connect tauri JS API with my test application but I'm getting the following error on a debug build in the browser console. 

JS error: `window.__TAURI_IPC__ is not a function.`

Then when I inspect what is available, I have the following

```
window.__TAURI_POST_MESSAGE__
(message) => window.ipc.postMessage(JSON.stringify(message)) 
```

I am using `"@tauri-apps/api": "1.1.0",` and 

```
[build-dependencies]
tauri-build = { version = "1.1.1", features = [] }

[dependencies]
serde_json = "1.0"
serde = { version = "1.0", features = ["derive"] }
tauri = { version = "1.1.1", features = ["shell-sidecar"] }
async-std = "1.12.0"
```

**Comment by FabianLars** [comment, marked as answer]
No there is no version mismatch, at least the core IPC stuff should be compatible across all 1.x versions.

> but I'm getting the following error on a debug build **in the browser console.**

Tauri's apis only work in actual Tauri windows. Often people could make it work with a combination of https://github.com/tauri-apps/tauri-invoke-http/ and https://github.com/tauri-apps/tauri-plugin-localhost/ but that has its own limitations, as-in you can't use it together with your actual dev server and of course using these plugins is not recommended in production at all.

**Reply by jourdain** [reply]
Since it is preferable to use `tauri://localhost` for serving the main content and that I only truly need a ws: with no http calls. 
I've made the change but now I'm getting some security error that I'm not sure I understand. I tried to play with the `security/csp` field within `tauri.conf.json` but with no luck so far.

I've started with 

```json
"security": {
      "csp": "default-src 'self'"
 },
```

and tried to expend to 

```json
"security": {
      "csp": "default-src 'unsafe-hashes' 'unsafe-inline' ws: http: 'self'"
 },
```

but I'm still getting the same error in tauri-window/Safari
<img width="912" alt="Screen Shot 2022-10-23 at 11 36 54" src="https://user-images.githubusercontent.com/294518/197406997-7a7c7f21-2e09-4504-b37a-afc7e6f99eea.png">

One specificity with my application is that the ws will provide guidance to the actual JS app (served by tauri) and dynamically load js and css to really fill the app content.

Thanks for all your guidance !

**Reply by FabianLars** [reply]
You'll need to play around with this setting https://tauri.app/v1/api/config/#securityconfig.dangerousdisableassetcspmodification because the browser/WebView will ignore unsafe-eval if you specify hashes or nonces in the csp config, which we do for the bundled assets

**Reply by jourdain** [reply]
Thanks to your guidance I was able to get it working with the following settings:

```json
    "security": {
      "csp": "default-src 'unsafe-eval' 'unsafe-inline' ws: 'self'",
      "dangerousDisableAssetCspModification": true
    },
```

Now the native menu properly trigger my Python callback via my websocket. This is awesome!

The full working example is available [here](https://github.com/Kitware/trame-tauri/tree/master/examples/menu). 

Thanks for your help!
